include/ginkgo/core/base/executor.hpp

```diff
@@ -276,43 +276,95 @@ class Operation {
-public:                                                                      \
-    void run(std::shared_ptr<const ::gko::_type> exec) const override        \
-    {                                                                        \
-        this->call(counts{}, exec);                                          \
-    }                                                                        \
-                                                                             \
-private:                                                                     \
-    template <int... Ns>                                                     \
-    void call(::gko::syn::value_list<int, Ns...>,                            \
-              std::shared_ptr<const ::gko::_type> &exec) const               \
-    {                                                                        \
-        ::gko::kernels::_namespace::_kernel(                                 \
-            exec, std::forward<Args>(std::get<Ns>(data))...);                \
-    }                                                                        \
-    static_assert(true,                                                      \
-                  "This assert is used to counter the false positive extra " \
-                  "semi-colon warnings")
 
-#define GKO_DETAIL_DEFINE_RUN_OVERLOAD(_type, _namespace, _kernel, ...)      \
-public:                                                                      \
-    void run(std::shared_ptr<const ::gko::_type> exec) const override        \
-    {                                                                        \
-        this->call(counts{}, exec);                                          \
-    }                                                                        \
-                                                                             \
-private:                                                                     \
-    template <int... Ns>                                                     \
-    void call(::gko::syn::value_list<int, Ns...>,                            \
-              std::shared_ptr<const ::gko::_type> &exec) const               \
-    {                                                                        \
-        ::gko::kernels::_namespace::_kernel(                                 \
-            exec, std::forward<Args>(std::get<Ns>(data))...);                \
-    }                                                                        \
-    static_assert(true,                                                      \
-                  "This assert is used to counter the false positive extra " \
-                  "semi-colon warnings")
+namespace detail {
+
+
+constexpr static int ref_exec_tag = 0;
+constexpr static int omp_exec_tag = 1;
+constexpr static int cuda_exec_tag = 2;
+constexpr static int hip_exec_tag = 3;
+constexpr static int dpcpp_exec_tag = 4;
```

Comment:
maybe use enum?
It is good to me because we do not need to consider the user-customized executor and it is not possible before, neither.

Reply:
Yes, totally! For some reason I didn't think of using an `enum`